File: 03/03.jl
timestamp_nano() = Int(round(time() * 1e9))

score(char)::Int = ifelse(char <= 'Z',
    char - 'A' + 27,
    char - 'a' + 1
)

part1(data)::Int = sum(map(
    score,
    # should only be one intersection anyways
    first(
        line[1:length(line)÷2] ∩ line[length(line)÷2+1:end]
    ) for line in data
))

part2(data)::Int = sum(map(
    score,
    first(
        data[i] ∩ data[i+1] ∩ data[i+2]
    ) for i in 1:3:length(data)
))

get_data(path::String) =
    open(path) do in_file
        split(read(in_file, String), "\n")
    end

start = timestamp_nano()

data = get_data("03/input.txt")

println("part1: ", part1(data))
println("part2: ", part2(data))

end_ = timestamp_nano()
println("time: ", round((end_ - start) / 1e6, digits=3), "ms")

File: 03/03.py
from timing_util import Timing


def get_data(content: str) -> list[str]:
    return content.splitlines()


def score(char: str) -> int:
    if (o := ord(char)) <= ord('Z'):
        return 27 + (o - ord('A'))
    else:
        return 1 + (o - ord('a'))


def part1(data: list[str]) -> int:
    s = 0
    for line in data:
        middle = len(line) // 2
        char = (set(line[:middle]) & set(line[middle:])).pop()

        s += score(char)
    return s


def part2(data: list[str]) -> int:
    s = 0
    for i in range(len(data) // 3):
        group = data[i * 3:(i + 1) * 3]
        common_badge = (set(group[0]) & set(group[1]) & set(group[2])).pop()

        s += score(common_badge)
    return s


if __name__ == '__main__':
    with Timing():
        with open('03/input.txt') as in_file:
            data = get_data(in_file.read())

        print(f'part1: {part1(data)}')
        print(f'part2: {part2(data)}')
